paper

```diff
@@ -0,0 +1,298 @@
+        printf '  * pc, paperclip       | Build paperclip jar. Need to apply patches first.\n'
+        printf '  * i, install          | Build and install paper into the local repo.\n'
+        printf '  * m, mcdev            | Setup decompiled sources for non-modified NMS files to be imported into an IDE.\n'
+        printf '  * u, up, upstream     | Updates the submodules used by Paper to their latest upstream versions.\n'
+        printf '  * upc, upstreamcommit | Creates the correctly-formatted upstream commit after updating upstream.\n'
+        printf '  * c, clean            | Removes all generated files.\n'
+        printf '  * t, testserver       | Run the test server. Accepts specifying argument to run "reobf", "shadow" or "dev" (default) type.\n'
+        printf '  * con, continue       | Shortcut command for running git am --continue, or git rebase --continue.\n'
+        printf '\n'
+        printf ' These commands require the setup command before use:\n'
+        printf '  * r, root             | Change directory to the root of the project.\n'
+        printf '  * a. api              | Move to the Paper-API directory.\n'
+        printf '  * s, server           | Move to the Paper-Server directory.\n'
+        printf '  * td, testdirectory   | Move to the test-server directory.\n'
+        printf '  * e, edit             | Use to edit a specific patch, give it the argument "server", "api" or "continue"\n'
+        printf '                        | respectively to edit the correct project. Use the argument "continue" after\n'
+        printf '                        | the changes have been made to finish and rebuild patches. Can be called from anywhere.\n'
+        printf '\n'
+        printf '  * setup               | Add an alias to %s to allow full functionality of this script. Run as:\n' "${paper_rcpath}"
+        printf '                        |     . ./paper setup\n'
+        printf "                        | After you run this command you'll be able to just run \"paper\" from anywhere.\n"
+        printf '                        | The default name for the resulting alias is "paper", you can give an argument to override\n'
+        printf '                        | this default, such as:\n'
+        printf '                        |     . ./paper setup example\n'
+        printf '                        | Which will allow you to run "example" instead.\n'
+  }
+
+  ###########################
+  ### CHECK PREREQUISITES ###
+  ###########################
+  local -r java_not_found_message="was not found - download the JDK by following https://paper.readthedocs.io/en/latest/java-update/index.html"
+  # TODO: Verify which ones are required
+  _paper_is_dep_available git
+  _paper_is_dep_available patch
+  _paper_is_dep_available mvn
+  _paper_is_dep_available curl
+  _paper_is_dep_available javac "${java_not_found_message}"
+  _paper_is_dep_available jar "${java_not_found_message}"
```

Comment:
not anymore

Reply:
I confirmed curl is only used for checkoutpr script, which isn't used by this script, so it's fine to remove. But you still need the JDK - `gradlew` will show a decent message when JRE is not installed, but when JDK is missing, it shows this exception:
```
> Task :Paper-API:compileJava FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':Paper-API:compileJava'.
> Cannot invoke "javax.tools.JavaCompiler.getStandardFileManager(javax.tools.DiagnosticListener, java.util.Locale, java.nio.charset.Charset)" because "this.delegate" is null

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
```
Maybe it's worth leaving them as means for JDK detection?

Reply:
I was thinking that the JDK missing might actually produce a nicer message, I mean, in part it kinda makes sense, the JavaCompiler is not there, so, probs JDK specific but, erm... normies. as said before, these scripts should generally serve as idiot checks but don't need to be fool proof, the notation of ensuring that specific binaries are on the path is pretty gross especially when we don't *need* them to be on the path itself is pretty meh, but, for the sake of idiot checks? eeeer... am not sure

Reply:
You don't need a JDK, no. If it's erroring then I messed something up, will check on it.

Reply:
Also (lol):

> eeeer... am not sure

I am sure. A properly set up project does not need a JDK. Gradle handles that: https://github.com/PaperMC/Paper/blob/0ad55267a387bf9cea3b6d0caeb843b3253cca65/build.gradle.kts#L12-L14

Reply:
Can't reproduce, builds fine without a JDK anywhere to be found.

Reply:
Reproduced on a fresh machine:
Order CX31 from Hetzner (cheapest 8GB machine), with Debian 10 and your SSH key.

Then run the following:
```sh
apt update
# Setup git
apt install git
git config --global user.email "you@example.com"
git config --global user.name "Your Name"
# Setup Java (paper documentation for Debian)
apt install apt-transport-https software-properties-common gnupg wget
wget -qO - https://adoptopenjdk.jfrog.io/adoptopenjdk/api/gpg/key/public | apt-key add -
add-apt-repository https://adoptopenjdk.jfrog.io/adoptopenjdk/deb/
apt update
apt install adoptopenjdk-16-hotspot-jre # Install JRE, not JDK
# Clone paper and build
git clone https://github.com/PaperMC/Paper.git
cd Paper
./gradlew applyPatches
./gradlew paperclipJar
```
Result of that last command:
```
root@paper-test:~/Paper# ./gradlew paperclipJar
> Task :Paper-API:compileJava FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':Paper-API:compileJava'.
> Cannot invoke "javax.tools.JavaCompiler.getStandardFileManager(javax.tools.DiagnosticListener, java.util.Locale, java.nio.charset.Charset)" because "this.delegate" is null

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 20s
18 actionable tasks: 5 executed, 13 up-to-date
```

It somehow needs to have a JDK (because it has to compile that gradle code from Kotlin, right?), but fails when I don't install full JDK.

I'm fine with removing it despite that.

Reply:
I did test with a fresh machine with only a JRE lol.

The failure is literally an NPE:
> Cannot invoke "javax.tools.JavaCompiler.getStandardFileManager(javax.tools.DiagnosticListener, java.util.Locale, java.nio.charset.Charset)" because "this.delegate" is null

I don't doubt the failure, but I'll follow your repro steps to figure out what's really going on.